ddl/dist_backfilling.go

```diff
@@ -0,0 +1,252 @@
+		return nil, errors.Trace(err)
+	}
+	bwCtx := &backfillWorkerContext{backfillWorkers: bCtxs, sessCtxs: make([]sessionctx.Context, 0, len(bCtxs))}
+	defer func() {
+		if err != nil {
+			bwCtx.close(d)
+		}
+	}()
+
+	for i := 0; i < len(bCtxs); i++ {
+		var se sessionctx.Context
+		se, err = d.sessPool.get()
+		if err != nil {
+			logutil.BgLogger().Error("[ddl] new backfill worker context, get a session failed", zap.Error(err))
+			return nil, errors.Trace(err)
+		}
+		err = initSessCtx(se, bfMeta.SQLMode, bfMeta.Location)
+		if err != nil {
+			logutil.BgLogger().Error("[ddl] new backfill worker context, init the session ctx failed", zap.Error(err))
+			return nil, errors.Trace(err)
+		}
+		bfCtx := newBackfillCtx(d.ddlCtx, 0, se, bfMeta.ReorgTp, schemaName, tbl)
+		var bf backfiller
+		bf, err = bfFunc(bfCtx)
+		if err != nil {
+			logutil.BgLogger().Error("[ddl] new backfill worker context, do bfFunc failed", zap.Error(err))
+			return nil, errors.Trace(err)
+		}
+		bwCtx.sessCtxs = append(bwCtx.sessCtxs, se)
+		bCtxs[i].backfiller = bf
+	}
+	return bwCtx, nil
+}
+
+func (bwCtx *backfillWorkerContext) GetContext() *backfillWorker {
+	bwCtx.mu.Lock()
+	// TODO: Special considerations are required if the number of consumers we get from the backfillWorkerPool is increased.
+	offset := bwCtx.currID % len(bwCtx.backfillWorkers)
+	// To prevent different workers from using the same session.
+	bw := bwCtx.backfillWorkers[offset]
+	logutil.BgLogger().Info("[ddl] backfill worker get context", zap.Int("workerCount", len(bwCtx.backfillWorkers)),
+		zap.Int("currID", bwCtx.currID), zap.Int("offset", offset), zap.Stringer("backfill worker", bw))
+	bwCtx.currID++
+	bwCtx.mu.Unlock()
+	return bw
+}
+
+func runBackfillJobs(d *ddl, sess *session, bJob *BackfillJob, jobCtx *JobContext) (table.Table, error) {
+	dbInfo, tbl, err := d.getTableByTxn(d.store, bJob.Meta.SchemaID, bJob.Meta.TableID)
+	if err != nil {
+		logutil.BgLogger().Warn("[ddl] runBackfillJobs gets table failed", zap.String("bfJob", bJob.AbbrStr()), zap.Error(err))
+		return nil, err
+	}
+
+	workerCnt := int(variable.GetDDLReorgWorkerCounter())
+	// TODO: Different worker using different newBackfillerFunc.
+	workerCtx, err := newAddIndexWorkerContext(d, dbInfo.Name, tbl, workerCnt, bJob, jobCtx)
+	if err != nil || workerCtx == nil {
+		logutil.BgLogger().Info("[ddl] new adding index worker context failed", zap.Reflect("workerCtx", workerCtx), zap.Error(err))
+		return nil, errors.Trace(err)
+	}
+	bwMgr := newBackfilWorkerManager(workerCtx)
+	d.backfillWorkerPool.SetConsumerFunc(func(task *reorgBackfillTask, _ int, bfWorker *backfillWorker) *backfillResult {
+		return bfWorker.runTask(task)
+	})
+	proFunc := func() ([]*reorgBackfillTask, error) {
+		// TODO: After BackfillJob replaces reorgBackfillTask, use backfiller's GetTasks instead of it.
+		return GetTasks(d.ddlCtx, sess, tbl, bJob.JobID, workerCnt*2)
```

Comment:
Why is `workerCnt * 2`? one to process and another one in waiting?

Reply:
Yes, pre-occupy the `workerCnt` tasks.